File: backend/app/models/roadmap.py
"""Roadmap feature: reusable templates + live per-student roadmaps.

Templates (roadmap_templates → template_stages → template_tasks → template_subtasks)
are authored by staff per country/degree. Assigning a template materializes a live
roadmap for a student (roadmaps → stages → roadmap_tasks → roadmap_subtasks), keyed
on students.id so staff control it from the CRM card and the student sees it in the
portal.
"""
import uuid
from datetime import datetime, date, timezone
import enum

from sqlalchemy import String, Integer, Text, Date, DateTime, Boolean, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class TaskPriority(str, enum.Enum):
    required = "required"
    recommended = "recommended"
    optional = "optional"


class TaskAudience(str, enum.Enum):
    applicant = "applicant"       # visible/actionable by the student
    coordinator = "coordinator"   # internal task for staff


class RoadmapItemStatus(str, enum.Enum):
    planned = "planned"
    in_progress = "in_progress"
    done = "done"


class RoadmapStatus(str, enum.Enum):
    active = "active"
    archived = "archived"


def _now() -> datetime:
    return datetime.now(timezone.utc)


# ---------------------------------------------------------------------------
# Templates
# ---------------------------------------------------------------------------
class RoadmapTemplate(Base):
    __tablename__ = "roadmap_templates"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(300))
    country_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    country_ref_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("country_reference.id", ondelete="SET NULL"), nullable=True
    )
    degree: Mapped[str] = mapped_column(String(40), default="bachelors")
    year: Mapped[int] = mapped_column(Integer)
    description: Mapped[str] = mapped_column(Text, default="")
    source_notion_db_id: Mapped[str | None] = mapped_column(String(80), nullable=True, unique=True, index=True)
    source_notion_title: Mapped[str | None] = mapped_column(String(500), nullable=True)
    source_notion_last_edited_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now)

    stages: Mapped[list["TemplateStage"]] = relationship(
        back_populates="template", cascade="all, delete-orphan", order_by="TemplateStage.position"
    )


class TemplateStage(Base):
    __tablename__ = "template_stages"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    template_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("roadmap_templates.id", ondelete="CASCADE"))
    name: Mapped[str] = mapped_column(String(300))
    description: Mapped[str] = mapped_column(Text, default="")
    position: Mapped[int] = mapped_column(Integer, default=0)

    template: Mapped["RoadmapTemplate"] = relationship(back_populates="stages")
    tasks: Mapped[list["TemplateTask"]] = relationship(
        back_populates="stage", cascade="all, delete-orphan", order_by="TemplateTask.position"
    )


class TemplateTask(Base):
    __tablename__ = "template_tasks"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    stage_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("template_stages.id", ondelete="CASCADE"))
    title: Mapped[str] = mapped_column(String(500))
    description: Mapped[str] = mapped_column(Text, default="")
    expected_result: Mapped[str] = mapped_column(Text, default="", server_default="")
    needs_document: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    needs_zoom: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    questionnaire_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    source_notion_page_id: Mapped[str | None] = mapped_column(String(80), nullable=True, unique=True, index=True)
    priority: Mapped[TaskPriority] = mapped_column(
        SAEnum(TaskPriority, name="task_priority"), default=TaskPriority.required
    )
    audience: Mapped[TaskAudience] = mapped_column(
        SAEnum(TaskAudience, name="task_audience"), default=TaskAudience.applicant
    )
    due_offset_days: Mapped[int | None] = mapped_column(Integer, nullable=True)
    position: Mapped[int] = mapped_column(Integer, default=0)

    stage: Mapped["TemplateStage"] = relationship(back_populates="tasks")
    subtasks: Mapped[list["TemplateSubtask"]] = relationship(
        back_populates="task", cascade="all, delete-orphan", order_by="TemplateSubtask.position"
    )


class TemplateSubtask(Base):
    __tablename__ = "template_subtasks"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    task_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("template_tasks.id", ondelete="CASCADE"))
    title: Mapped[str] = mapped_column(String(500))
    position: Mapped[int] = mapped_column(Integer, default=0)
    source_notion_page_id: Mapped[str | None] = mapped_column(String(80), nullable=True, unique=True, index=True)

    task: Mapped["TemplateTask"] = relationship(back_populates="subtasks")


# ---------------------------------------------------------------------------
# Live roadmaps (per student)
# ---------------------------------------------------------------------------
class Roadmap(Base):
    __tablename__ = "roadmaps"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    student_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("students.id", ondelete="CASCADE"), index=True)
    mentor_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    template_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("roadmap_templates.id", ondelete="SET NULL"), nullable=True
    )
    name: Mapped[str] = mapped_column(String(300))
    country_name: Mapped[str | None] = mapped_column(String(200), nullable=True)
    degree: Mapped[str] = mapped_column(String(40), default="bachelors")
    year: Mapped[int] = mapped_column(Integer)
    status: Mapped[RoadmapStatus] = mapped_column(
        SAEnum(RoadmapStatus, name="roadmap_status"), default=RoadmapStatus.active
    )
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now)

    stages: Mapped[list["Stage"]] = relationship(
        back_populates="roadmap", cascade="all, delete-orphan", order_by="Stage.position"
    )


class Stage(Base):
    __tablename__ = "stages"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    roadmap_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("roadmaps.id", ondelete="CASCADE"), index=True)
    name: Mapped[str] = mapped_column(String(300))
    description: Mapped[str] = mapped_column(Text, default="")
    position: Mapped[int] = mapped_column(Integer, default=0)
    status: Mapped[RoadmapItemStatus] = mapped_column(
        SAEnum(RoadmapItemStatus, name="roadmap_item_status"), default=RoadmapItemStatus.planned
    )

    roadmap: Mapped["Roadmap"] = relationship(back_populates="stages")
    tasks: Mapped[list["RoadmapTask"]] = relationship(
        back_populates="stage", cascade="all, delete-orphan", order_by="RoadmapTask.position"
    )


class RoadmapTask(Base):
    __tablename__ = "roadmap_tasks"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    stage_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("stages.id", ondelete="CASCADE"), index=True)
    roadmap_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("roadmaps.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(500))
    description: Mapped[str] = mapped_column(Text, default="")
    expected_result: Mapped[str] = mapped_column(Text, default="", server_default="")
    needs_document: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    needs_zoom: Mapped[bool] = mapped_column(Boolean, default=False, server_default="false")
    questionnaire_url: Mapped[str | None] = mapped_column(String(2048), nullable=True)
    assignee_id: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    created_by: Mapped[uuid.UUID | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    priority: Mapped[TaskPriority] = mapped_column(
        SAEnum(TaskPriority, name="task_priority"), default=TaskPriority.required
    )
    audience: Mapped[TaskAudience] = mapped_column(
        SAEnum(TaskAudience, name="task_audience"), default=TaskAudience.applicant
    )
    status: Mapped[RoadmapItemStatus] = mapped_column(
        SAEnum(RoadmapItemStatus, name="roadmap_item_status"), default=RoadmapItemStatus.planned
    )
    due_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    position: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=_now, onupdate=_now)

    stage: Mapped["Stage"] = relationship(back_populates="tasks")
    subtasks: Mapped[list["RoadmapSubtask"]] = relationship(
        back_populates="task", cascade="all, delete-orphan", order_by="RoadmapSubtask.position"
    )


class RoadmapSubtask(Base):
    __tablename__ = "roadmap_subtasks"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    task_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("roadmap_tasks.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(500))
    is_done: Mapped[bool] = mapped_column(Boolean, default=False)
    position: Mapped[int] = mapped_column(Integer, default=0)

    task: Mapped["RoadmapTask"] = relationship(back_populates="subtasks")

File: frontend/src/api/roadmap.ts
import apiClient from './client'

export type ItemStatus = 'planned' | 'in_progress' | 'done'
export type Priority = 'required' | 'recommended' | 'optional'
export type Audience = 'applicant' | 'coordinator'

export interface RoadmapSubtask {
  id: string
  title: string
  is_done: boolean
  position: number
}

export interface RoadmapTask {
  id: string
  stage_id: string
  roadmap_id: string
  title: string
  description: string
  expected_result: string
  needs_document: boolean
  needs_zoom: boolean
  questionnaire_url: string | null
  priority: Priority
  audience: Audience
  status: ItemStatus
  due_date: string | null
  position: number
  subtasks: RoadmapSubtask[]
}

export interface RoadmapStage {
  id: string
  roadmap_id: string
  name: string
  description: string
  position: number
  status: ItemStatus
  tasks: RoadmapTask[]
}

export interface Roadmap {
  id: string
  student_id: string
  mentor_id: string | null
  mentor_name?: string | null
  template_id: string | null
  name: string
  country_name: string | null
  country_flag_emoji?: string
  country_flag_url?: string
  degree: string
  year: number
  status: string
  created_at: string
  stages: RoadmapStage[]
}

// ---- templates ----
export interface TemplateListItem {
  id: string
  name: string
  country_name: string | null
  degree: string
  year: number
  stage_count: number
  task_count: number
  source_notion_db_id?: string | null
  source_notion_title?: string | null
  source_notion_last_edited_at?: string | null
}

export interface TemplateSubtask {
  id: string
  title: string
  position: number
  source_notion_page_id?: string | null
}
export interface TemplateTask {
  id: string
  title: string
  description: string
  expected_result: string
  needs_document: boolean
  needs_zoom: boolean
  questionnaire_url: string | null
  source_notion_page_id?: string | null
  priority: Priority
  audience: Audience
  due_offset_days: number | null
  position: number
  subtasks: TemplateSubtask[]
}
export interface TemplateStage {
  id: string
  name: string
  description: string
  position: number
  tasks: TemplateTask[]
}
export interface RoadmapTemplate {
  id: string
  name: string
  country_name: string | null
  degree: string
  year: number
  description: string
  source_notion_db_id?: string | null
  source_notion_title?: string | null
  source_notion_last_edited_at?: string | null
  created_at: string
  stages: TemplateStage[]
}

// structure input (create / put-structure)
export interface SubtaskInput {
  title: string
  source_notion_page_id?: string | null
}
export interface TaskInput {
  title: string
  description?: string
  expected_result?: string
  needs_document?: boolean
  needs_zoom?: boolean
  questionnaire_url?: string | null
  source_notion_page_id?: string | null
  priority?: Priority
  audience?: Audience
  due_offset_days?: number | null
  subtasks?: SubtaskInput[]
}
export interface StageInput {
  name: string
  description?: string
  tasks?: TaskInput[]
}
export interface TemplateCreateInput {
  name: string
  country_name?: string | null
  degree?: string
  year: number
  description?: string
  source_notion_db_id?: string | null
  source_notion_title?: string | null
  source_notion_last_edited_at?: string | null
  stages?: StageInput[]
}

export interface NotionRoadmapImportRequest {
  dry_run?: boolean
  discover_only?: boolean
  skip_subtasks?: boolean
  only?: string | null
}

export interface NotionRoadmapImportJob {
  job_id: string
  status: 'running' | 'done' | 'failed'
  started_at: string
  finished_at: string | null
  request: NotionRoadmapImportRequest
  progress: {
    message?: string
    template_index?: number
    template_total?: number
    task_index?: number
    task_total?: number
    title?: string
  }
  events: Array<Record<string, unknown>>
  result: null | {
    ok: boolean
    mode: string
    found: number
    created: number
    updated: number
    tasks: number
    subtasks: number
    templates: Array<{
      title: string
      label: string
      source_notion_db_id: string
      action: string
      stages: number
      tasks: number
      subtasks: number
    }>
    error: string | null
  }
  error: string | null
}

export interface FlatTask {
  id: string
  stage_id: string
  roadmap_id: string
  stage_name: string
  stage_position: number
  title: string
  description: string
  expected_result: string
  needs_document: boolean
  needs_zoom: boolean
  questionnaire_url: string | null
  priority: Priority
  audience: Audience
  status: ItemStatus
  due_date: string | null
  position: number
  subtasks: RoadmapSubtask[]
}

const data = <T>(p: Promise<{ data: T }>) => p.then((r) => r.data)

export const roadmapApi = {
  // live roadmap
  myRoadmap: () => data<Roadmap | null>(apiClient.get('/portal/roadmap')),
  studentRoadmap: (studentId: string) =>
    data<Roadmap | null>(apiClient.get(`/students/${studentId}/roadmap`)),

  myTasks: () => data<FlatTask[]>(apiClient.get('/portal/tasks')),
  studentTasks: (studentId: string) =>
    data<FlatTask[]>(apiClient.get(`/students/${studentId}/tasks`)),

  updateStage: (stageId: string, body: { status?: ItemStatus; name?: string; description?: string }) =>
    data<Roadmap>(apiClient.patch(`/stages/${stageId}`, body)),
  createTask: (body: {
    stage_id: string
    title: string
    description?: string
    expected_result?: string
    needs_document?: boolean
    needs_zoom?: boolean
    questionnaire_url?: string | null
    priority?: Priority
    audience?: Audience
    due_date?: string | null
  }) =>
    data<Roadmap>(apiClient.post('/roadmap-tasks', body)),
  updateTask: (taskId: string, body: Partial<{
    title: string
    description: string
    expected_result: string
    needs_document: boolean
    needs_zoom: boolean
    questionnaire_url: string | null
    status: ItemStatus
    priority: Priority
    audience: Audience
    due_date: string | null
  }>) =>
    data<Roadmap>(apiClient.patch(`/roadmap-tasks/${taskId}`, body)),
  deleteTask: (taskId: string) => apiClient.delete(`/roadmap-tasks/${taskId}`),
  createSubtask: (taskId: string, title: string) =>
    data<Roadmap>(apiClient.post(`/roadmap-tasks/${taskId}/subtasks`, { title })),
  updateSubtask: (subtaskId: string, body: { is_done?: boolean; title?: string }) =>
    data<Roadmap>(apiClient.patch(`/roadmap-subtasks/${subtaskId}`, body)),
  deleteSubtask: (subtaskId: string) => apiClient.delete(`/roadmap-subtasks/${subtaskId}`),

  // templates
  listTemplates: () => data<TemplateListItem[]>(apiClient.get('/roadmap-templates')),
  getTemplate: (id: string) => data<RoadmapTemplate>(apiClient.get(`/roadmap-templates/${id}`)),
  createTemplate: (body: TemplateCreateInput) =>
    data<RoadmapTemplate>(apiClient.post('/roadmap-templates', body)),
  updateTemplate: (id: string, body: Partial<TemplateCreateInput>) =>
    data<RoadmapTemplate>(apiClient.patch(`/roadmap-templates/${id}`, body)),
  deleteTemplate: (id: string) => apiClient.delete(`/roadmap-templates/${id}`),
  putStructure: (id: string, stages: StageInput[]) =>
    data<RoadmapTemplate>(apiClient.put(`/roadmap-templates/${id}/structure`, { stages })),
  assign: (templateId: string, studentId: string, opts?: { mentor_id?: string; name?: string }) =>
    data<Roadmap>(apiClient.post(`/roadmap-templates/${templateId}/assign`, { student_id: studentId, ...opts })),
  startNotionImport: (body: NotionRoadmapImportRequest) =>
    data<NotionRoadmapImportJob>(apiClient.post('/roadmap-templates/import/notion', body)),
  notionImportJob: (jobId: string) =>
    data<NotionRoadmapImportJob>(apiClient.get(`/roadmap-templates/import/notion/${jobId}`)),
  notionImportJobs: () => data<NotionRoadmapImportJob[]>(apiClient.get('/roadmap-templates/import/notion')),
}
